supporter-product-data/src/main/scala/com/gu/lambdas/ProcessSupporterRatePlanItemLambda.scala

```diff
@@ -0,0 +1,132 @@
+    Future
+      .sequence(input.Records.map { record =>
+        val subscription = decode[SupporterRatePlanItem](record.body)
+        subscription match {
+          case Right(item) => supporterRatePlanItemProcessor.processItem(item)
+          case _ =>
+            SafeLogger.warn(s"Couldn't decode a SupporterRatePlanItem with body: ${record.body}")
+            Future.successful(()) // This should never happen so I don't think it's worth alerting on
+        }
+      })
+      .map(_ => ())
+  }
+}
+
+object ProcessSupporterRatePlanItemLambda {
+  def getSupporterRatePlanItemProcessor() = {
+    val stage = StageConstructors.fromEnvironment
+    val dynamoService = SupporterDataDynamoService(stage)
+    val alarmService = AlarmService(stage)
+    val futureProcessor = ConfigService(stage).load.map { config =>
+      new SupporterRatePlanItemProcessor(
+        dynamoService,
+        alarmService,
+        new ContributionProcessor(stage, config),
+      )
+    }
+    Await.result(futureProcessor, 10.seconds)
+  }
+}
+
+class ContributionProcessor(stage: Stage, config: ZuoraQuerierConfig) {
+  val contributionIds = ContributionIds.forStage(stage)
+  lazy val zuoraService = new ZuoraSubscriptionService(config, configurableFutureRunner(60.seconds))
+
+  def isRecurringContribution(supporterRatePlanItem: SupporterRatePlanItem) =
+    contributionIds.contains(supporterRatePlanItem.productRatePlanId)
+
+  def addAmountIfContribution(supporterRatePlanItem: SupporterRatePlanItem) =
+    if (isRecurringContribution(supporterRatePlanItem)) {
+      SafeLogger.info(s"Supporter ${supporterRatePlanItem.identityId} is a recurring contributor")
+      fetchContributionAmountFromZuora(supporterRatePlanItem)
+    } else
+      Future.successful(supporterRatePlanItem)
+
+  def fetchContributionAmountFromZuora(supporterRatePlanItem: SupporterRatePlanItem) =
+    zuoraService
+      .getSubscription(supporterRatePlanItem.subscriptionName)
+      .map { sub =>
+        SafeLogger.info(
+          s"Contribution amount for supporter ${supporterRatePlanItem.identityId} is ${sub.contributionAmount}",
+        )
+        supporterRatePlanItem.copy(contributionAmount = sub.contributionAmount)
+      }
+
+}
+
+class SupporterRatePlanItemProcessor(
+    dynamoService: SupporterDataDynamoService,
+    alarmService: AlarmService,
+    contributionProcessor: ContributionProcessor,
+) {
+
+  def processItem(supporterRatePlanItem: SupporterRatePlanItem) =
+    contributionProcessor
+      .addAmountIfContribution(supporterRatePlanItem)
+      .flatMap(writeItemToDynamo)
+
+  def writeItemToDynamo(supporterRatePlanItem: SupporterRatePlanItem) = {
+    SafeLogger.info(s"Attempting to write item to Dynamo - ${supporterRatePlanItem.asJson.noSpaces}")
+    try {
+      dynamoService.writeItem(supporterRatePlanItem).map { _ =>
+        SafeLogger.info(s"Successfully wrote item to Dynamo - ${supporterRatePlanItem.asJson.noSpaces}")
+        ()
+      }
+    } catch {
```

Comment:
this won't do much as writeItem returns a Future - we need to be looking out for a Failed future which I think will happen at the handlerFuture level.  Ideally we can catch all exceptions at the lowest level, and deal in Futures and Trys everywhere else.